Update project docs (#571)
elastic/integrations: docs/developer_workflow_design_build_test_integration.md

# Developer workflow: build and test the integration

## Prerequisites

> * `elastic-package` (builder tool) installed - follow the [Getting started](https://github.com/elastic/elastic-package#getting-started) guide to install the tool. 

Reviewer: ```suggestion
* `elastic-package` (builder tool) installed - follow the [Getting Started Guide](https://github.com/elastic/elastic-package#getting-started) guide to install the tool. 
```

* If you don't understand the `elastic-package` command or would like to learn more about it, try the `help` switch, e.g. `elastic-package stack up --help`.

## Steps

I assume that you've selected a product or service from which you'd like to collect logs and metrics. If this is your

> first interaction with Integrations or packages, I suggest to review existing Integrations in the Fleet UI.

Reviewer: > I suggest to review existing Integrations in the Fleet UI.

The motivation behind this suggestion is a bit unclear to me. Maybe add something about how this suggestion would benefit the reader?

Author: I added more information what to look at.

Let's bring up the Elastic stack:

```bash
elastic-package stack up -v -d
```

Navigate to the [Integrations](http://localhost:5601/app/fleet#/integration) page, try to add an integration to the default
policy, review forms. The good candidate to start your journey are "Nginx", "Apache", "Nats".

Once you selected the donor for your integration, copy the package source, e.g.:

```bash
cd packages
cp -r nginx new_package
```

> where `new_package` is the name of your integration.

Reviewer: Some day it would be nice if we had `elastic-package init` (or similar) that setup a basic package structure for new package developers, giving them an option to copy from a donor package. 🙂

Author: https://github.com/elastic/elastic-package/issues/3

Review all resources, remove unnecessary ones, adjusts manifests, create new data streams.

### Build

Now, it's the moment to build the package:

```bash
elastic-package build
```

... and recycle the package-registry Docker container (run from inside of the integration directory):

```bash
elastic-package stack up --services package-registry
```

Once the container is recycled, you can refresh the Fleet UI and Kibana will pick up updated packages.

### Lint

You can verify if the package is aligned with the package-spec using:

```bash
elastic-package lint
```

The command will show potential problems with linting and give you a suggestion on how to fix it.

### Format

You can format the package contents (JSON, YML files) with:

```bash
elastic-package format
```

### Export resources

If you're working on Kibana dashboards and would like to export them to local directories, run the following command
(run from inside of the integration directory):

```bash
elastic-package export
```

... and follow TUI steps (dashboard selection).

### Test

> The `elastic-package` tool supports multiple levels of testing - pipeline, system, assets. Follow up on the specific topic

Reviewer: ```suggestion
The `elastic-package` tool supports multiple types of tests - pipeline, system, assets. Follow up on the specific topic
```

using the tool's [documentation](https://github.com/elastic/elastic-package/tree/master/docs/howto).

### Remember to bump up the version

When the PR is merged, the CI will kick off a build job for the master branch, which can release your integration to
the package-storage. It means that it will open a PR to the [Package Storage/snapshot](https://github.com/elastic/package-storage/tree/snapshot/packages) with
the built integration if only the package version doesn't already exist in the storage (hasn't been released yet).

> Long story short - if you want your changes in the integration to be released, remember to bump up the package version.

Reviewer: ```suggestion
Long story short - when you are ready for your changes in the integration to be released, remember to bump up the package version.
```

> ### Open a PR

Reviewer: I think this section makes sense before the version bump section?

Author: I changed the order of "Open a PR" and "Remember to bump up the version".
